zerver/lib/actions.py

```diff
@@ -4562,8 +4595,27 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
             # Remove duplicates by excluding the id of users already in users_to_be_notified list.
             # This is the case where a user both has a UserMessage row and is a current Subscriber
             subscribers = subscribers.exclude(user_profile_id__in=[um.user_profile_id for um in ums])
+
+            if new_stream is not None:
+                assert guest_subs is not None
+                # Exclude guest users who are not subscribed to the new stream from receing this event.
+                subscribers = subscribers.exclude(user_profile_id__in=[sub.user_profile_id for sub in guest_subs])
```

Comment:
I think we'll end up with a better query if we combine this with the other `__in` exclusion.